File: Scripts/Combatant.cs
using Godot;
using System;

public class Combatant : KinematicBody
{

    //PHYSICS VARIABLES--------------------------------
    [Export]
    public float Gravity = -20f;
    [Export]
    public float MaxSpeed = 8.0f;
    [Export]
    public float Accel = 4.5f;
    [Export]
    public float Deaccel = 16.0f;
    [Export]
    public float MaxSlopeAngle = 50.0f;
    //GAMEPLAY VARIABLES-------------------------------------
    [Export]
    public int HP = 100;

    //COMPONENT VARIABLES------------------------------------
    protected Spatial head;
    protected Vector3 vel = new Vector3();
    protected Vector3 dir = new Vector3();
    
    // Called when the node enters the scene tree for the first time.
    public override void _Ready()
    {
        
    }

//  // Called every frame. 'delta' is the elapsed time since the previous frame.
    public override void _PhysicsProcess(float delta)
    {
        ProcessMovement(delta);
    }
    protected virtual void ProcessMovement(float delta)
    {
        dir.y = 0;
        dir = dir.Normalized();

        vel.y += delta * Gravity;
        
        Vector3 hvel = vel;
        hvel.y = 0;

        Vector3 target = dir;

        target *= MaxSpeed;

        float accel;
        if (dir.Dot(hvel) > 0)
            accel = Accel;
        else
            accel = Deaccel;

        hvel = hvel.LinearInterpolate(target, accel * delta);
        vel.x = hvel.x;
        vel.z = hvel.z;
        vel = MoveAndSlide(vel, new Vector3(0, 1, 0), false, 4, Mathf.Deg2Rad(MaxSlopeAngle));
    }
    public virtual void UpdateHealth(int damage)
    {
        GD.Print("owie" + damage.ToString());
        //TODO: hit sound
        HP -= damage;
    }
}

File: src/Combatant.cpp
#include "Combatant.h"

using namespace godot;

void Combatant::_register_methods() {
    register_property("HP", &Combatant::HP, 100);
    register_method("_process", &Combatant::_process);
    register_method("_physics_process", &Combatant::_physics_process);
    register_method("process_movement", &Combatant::process_movement);
    register_method("update_health", &Combatant::update_health);
    register_property("HP", &Combatant::HP, 100);
    register_property("Gravity", &Combatant::gravity, -20.0f);
    register_property("Max Speed", &Combatant::maxSpeed, 8.0f);
    register_property("Acceleration", &Combatant::accel, 4.5f);
    register_property("DeAcceleration", &Combatant::deAccel, 16.0f);
    register_property("Max Slope Angle", &Combatant::maxSlopeAngle, 50.0f);
}

Combatant::Combatant() {
    
}

Combatant::~Combatant() {
    // add your cleanup here
}

void Combatant::_init() {
    _head = nullptr;
    /* //wait do I even need to initialize variables here if they're registered up there?
    gravity = -20.0f;
    maxSpeed = 8.0f;
    accel = 4.5f;
    deAccel = 16.0f;
    maxSlopeAngle = 50.0f;
    HP = 100;
    */
}

void Combatant::_process(const double p_delta) {
    
}
void Combatant::_physics_process(const double p_delta) {
    process_movement(p_delta);
}
void Combatant::process_movement(const double p_delta) {
    dir.y = 0;
    dir = dir.normalized();

    vel.y += p_delta * gravity;

    Vector3 hvel = vel;
    hvel.y = 0;

    Vector3 target = dir;

    target *= maxSpeed;
    float accel;
    if (dir.dot(hvel) > 0)
        accel = accel;
    else
        accel = deAccel;

    hvel = hvel.linear_interpolate(target, accel * p_delta);
    vel.x = hvel.x;
    vel.z = hvel.z;

    hvel = hvel.linear_interpolate(target, (dir.dot(hvel) > 0) ? accel : deAccel * p_delta);
    
    vel.x = hvel.x;
    vel.z = hvel.z;
    vel = move_and_slide(vel, Vector3::UP, false, 4, Math::deg2rad(maxSlopeAngle));
}
void Combatant::update_health(const int delta) {
    HP -= delta;
}